File: src/linksearch/platforms/youtube_shorts_ytdlp.py
"""YouTube Shorts discovery via yt-dlp using YouTube's Shorts search filter (real Shorts)."""

from __future__ import annotations

import logging
import os
import shutil
import subprocess
import sys
from typing import TYPE_CHECKING
from urllib.parse import urlencode, urlparse

from linksearch.models import CandidateLink, ProductInput

if TYPE_CHECKING:
    from linksearch.config import Settings

logger = logging.getLogger(__name__)

# YouTube web search: Type = Shorts (not plain "ytsearch" + keyword — that returns mixed videos).
# Same filter as opening Results with the Shorts chip; see yt-dlp YoutubeSearchURLIE / issue #16010.
_SHORTS_SEARCH_SP = "EgIQCQ=="


def _shorts_results_url(search_query: str) -> str:
    q = search_query.strip()
    if not q:
        return ""
    return "https://www.youtube.com/results?" + urlencode(
        {"search_query": q, "sp": _SHORTS_SEARCH_SP}
    )


def _shorts_watch_url(url: str) -> str | None:
    """Normalize to https://www.youtube.com/shorts/VIDEO_ID when possible."""
    if not url or not isinstance(url, str):
        return None
    u = url.strip()
    if "/shorts/" in u.lower():
        try:
            p = urlparse(u)
            path = p.path or ""
            if "/shorts/" in path.lower():
                vid = path.split("/shorts/", 1)[-1].strip("/").split("/")[0]
                if len(vid) == 11:
                    return f"https://www.youtube.com/shorts/{vid}"
        except Exception:
            pass
        return u.split("&")[0]
    return None


def _env_date_range() -> tuple[str | None, str | None]:
    """CLI / server env: YOUTUBE_SHORTS_DATE_AFTER, YOUTUBE_SHORTS_DATE_BEFORE (YYYY-MM-DD)."""
    a = os.environ.get("YOUTUBE_SHORTS_DATE_AFTER", "").strip()
    b = os.environ.get("YOUTUBE_SHORTS_DATE_BEFORE", "").strip()
    if not a and not b:
        return None, None
    return (a or None, b or None)


def _iso_to_ymd(iso: str) -> str:
    return iso.replace("-", "")[:8]


def _upload_date_ydl(short_url: str, timeout: float) -> str | None:
    """Return upload_date as YYYYMMDD via yt-dlp --print."""
    tail = [
        short_url,
        "--skip-download",
        "--print",
        "%(upload_date)s",
        "--no-warnings",
        "--quiet",
    ]
    candidates: list[list[str]] = []
    if shutil.which("yt-dlp"):
        candidates.append(["yt-dlp", *tail])
    candidates.append([sys.executable, "-m", "yt_dlp", *tail])
    for cmd in candidates:
        try:
            r = subprocess.run(
                cmd,
                capture_output=True,
                text=True,
                timeout=max(15.0, timeout),
            )
            line = (r.stdout or "").strip()
            if len(line) == 8 and line.isdigit():
                return line
        except Exception:
            continue
    return None


def search_youtube_shorts_ytdlp_sync(
    settings: "Settings",
    product: ProductInput,
    queries: list[str],
) -> list[CandidateLink]:
    """Search YouTube Shorts only; emit https://www.youtube.com/shorts/VIDEO_ID links."""
    if not settings.youtube_use_ytdlp:
        return []
    try:
        import yt_dlp
    except ImportError:
        logger.warning("yt-dlp is not installed; skip YouTube Shorts search.")
        return []

    cap = max(1, settings.max_results_per_platform)
    base = " ".join(queries[:3]) if queries else product.primary_query()
    base = base.strip()[:200]
    if not base:
        return []
    q_text = base
    page_url = _shorts_results_url(q_text)
    if not page_url:
        return []

    d_after_raw, d_before_raw = _env_date_range()
    has_date = bool(d_after_raw or d_before_raw)
    max_collect = min(max(cap * 10, 24), 50) if has_date else cap
    after_ymd = _iso_to_ymd(d_after_raw) if d_after_raw else None
    before_ymd = _iso_to_ymd(d_before_raw) if d_before_raw else None

    ydl_opts: dict = {
        "quiet": True,
        "no_warnings": True,
        "extract_flat": "in_playlist",
        "skip_download": True,
        "playlistend": max_collect,
        "socket_timeout": min(60.0, max(15.0, settings.request_timeout_seconds + 15)),
    }
    try:
        with yt_dlp.YoutubeDL(ydl_opts) as ydl:
            info = ydl.extract_info(page_url, download=False)
    except Exception as e:
        logger.warning("yt-dlp YouTube Shorts search failed: %s", e)
        return []

    entries: list = []
    if info is None:
        return []
    if info.get("_type") == "playlist":
        entries = [e for e in (info.get("entries") or []) if e]
    else:
        entries = [info]

    out: list[CandidateLink] = []
    seen: set[str] = set()
    brand = product.normalized_brand()
    sku = product.sku.strip()
    pname = product.product_name.strip()

    for ent in entries:
        if not isinstance(ent, dict):
            continue
        vid = ent.get("id")
        vid_s = str(vid).strip() if vid is not None else ""
        title = str(ent.get("title") or "")
        raw_url = ent.get("url")
        u = _shorts_watch_url(str(raw_url)) if raw_url else None
        if not u and len(vid_s) == 11:
            u = f"https://www.youtube.com/shorts/{vid_s}"
        if not u:
            continue
        if u in seen:
            continue
        seen.add(u)
        out.append(
            CandidateLink(
                media="YoutubeShorts",
                brand=brand,
                url=u,
                sku=sku,
                product_name=pname,
                title=title,
                snippet=str(ent.get("description") or "")[:500],
                source_query=q_text,
            )
        )
        if len(out) >= max_collect:
            break

    if not has_date:
        return out[:cap]

    tout = min(60.0, max(15.0, settings.request_timeout_seconds + 15))
    filtered: list[CandidateLink] = []
    for c in out:
        ud = _upload_date_ydl(c.url, tout)
        if not ud:
            continue
        if after_ymd and ud < after_ymd:
            continue
        if before_ymd and ud > before_ymd:
            continue
        filtered.append(c)
        if len(filtered) >= cap:
            break
    return filtered

File: frontend/src/lib/platforms/youtubeShortsYtdlp.ts
import { execFile } from "node:child_process";
import { promisify } from "node:util";

import type { CandidateLink, ProductInput, SearchSettings } from "../types";
import { primaryQuery } from "../scoring";

const execFileAsync = promisify(execFile);

/** YouTube web: Type = Shorts (matches Shorts chip on /results). */
const SHORTS_SEARCH_SP = "EgIQCQ==";

export type ShortsDateRange = {
  /** Inclusive, YYYY-MM-DD */
  after?: string;
  /** Inclusive, YYYY-MM-DD */
  before?: string;
};

function shortsResultsUrl(searchQuery: string): string {
  const u = new URL("https://www.youtube.com/results");
  u.searchParams.set("search_query", searchQuery.trim());
  u.searchParams.set("sp", SHORTS_SEARCH_SP);
  return u.toString();
}

function normalizeShortsUrl(raw: string | undefined, id: string): string | null {
  if (raw && typeof raw === "string" && raw.includes("/shorts/")) {
    try {
      const p = new URL(raw).pathname.split("/shorts/")[1]?.split("/")[0];
      if (p && p.length === 11) return `https://www.youtube.com/shorts/${p}`;
    } catch {
      /* fall through */
    }
  }
  if (id.length === 11) return `https://www.youtube.com/shorts/${id}`;
  return null;
}

function parsePlaylistJson(raw: string): CandidateLink[] {
  const info = JSON.parse(raw) as {
    _type?: string;
    entries?: unknown[];
    id?: string;
    title?: string;
    url?: string;
    description?: string;
  };
  let entries: unknown[];
  if (info._type === "playlist") {
    entries = (info.entries ?? []).filter(Boolean);
  } else if (info.id || info.url) {
    entries = [info];
  } else {
    entries = [];
  }
  const out: CandidateLink[] = [];
  for (const ent of entries) {
    if (!ent || typeof ent !== "object") continue;
    const e = ent as Record<string, unknown>;
    const id = typeof e.id === "string" ? e.id : "";
    const title = String(e.title ?? "");
    const rawUrl = typeof e.url === "string" ? e.url : undefined;
    const url = normalizeShortsUrl(rawUrl, id);
    if (!url) continue;
    out.push({
      media: "YoutubeShorts",
      brand: "",
      url,
      sku: "",
      productName: "",
      title,
      snippet: String(e.description ?? "").slice(0, 500),
      score: 0,
    });
  }
  return out;
}

async function runYtdlpStdout(args: string[]): Promise<string | null> {
  const attempts: { cmd: string; argv: string[] }[] = [
    { cmd: process.env.YTDLP_BIN || "yt-dlp", argv: args },
    { cmd: "python", argv: ["-m", "yt_dlp", ...args] },
    { cmd: "py", argv: ["-m", "yt_dlp", ...args] },
  ];
  for (const { cmd, argv } of attempts) {
    try {
      const { stdout } = await execFileAsync(cmd, argv, {
        timeout: 90_000,
        maxBuffer: 25 * 1024 * 1024,
        windowsHide: true,
      });
      return stdout.toString();
    } catch {
      /* try next */
    }
  }
  return null;
}

/** yt-dlp upload_date as YYYYMMDD */
async function fetchUploadDateYmd(shortUrl: string): Promise<string | null> {
  const args = [
    shortUrl,
    "--skip-download",
    "--print",
    "%(upload_date)s",
    "--no-warnings",
    "--quiet",
  ];
  const out = await runYtdlpStdout(args);
  const line = out?.trim();
  if (!line || !/^\d{8}$/.test(line)) return null;
  return line;
}

function isoToYmd(iso: string): string {
  return iso.replace(/-/g, "");
}

async function filterRowsByUploadDate(
  rows: CandidateLink[],
  range: ShortsDateRange,
  cap: number
): Promise<CandidateLink[]> {
  const after = range.after ? isoToYmd(range.after) : null;
  const before = range.before ? isoToYmd(range.before) : null;
  if (!after && !before) return rows.slice(0, cap);

  const out: CandidateLink[] = [];
  for (const row of rows) {
    const ud = await fetchUploadDateYmd(row.url);
    if (!ud) continue;
    if (after && ud < after) continue;
    if (before && ud > before) continue;
    out.push(row);
    if (out.length >= cap) break;
  }
  return out;
}

/**
 * YouTube Shorts only: /results with Shorts type filter (not generic ytsearch).
 * Optional upload-date range: fetches per-video metadata (yt-dlp) so results match the window.
 */
export async function searchYoutubeShortsYtdlp(
  settings: SearchSettings,
  product: ProductInput,
  queries: string[],
  dateRange?: ShortsDateRange
): Promise<CandidateLink[]> {
  if (!settings.youtubeUseYtdlp) return [];
  const cap = settings.maxResultsPerPlatform;
  const base = (queries.length
    ? queries.slice(0, 3).join(" ")
    : primaryQuery(product)
  )
    .trim()
    .slice(0, 200);
  if (!base) return [];
  const qText = base;
  const pageUrl = shortsResultsUrl(qText);

  const hasDate =
    Boolean(dateRange?.after?.trim()) || Boolean(dateRange?.before?.trim());
  const playlistEnd = hasDate ? Math.min(Math.max(cap * 10, 24), 50) : cap;

  const args = [
    pageUrl,
    "-J",
    "--flat-playlist",
    "--no-download",
    "--no-warnings",
    "--quiet",
    "--playlist-end",
    String(playlistEnd),
  ];
  const stdout = await runYtdlpStdout(args);
  if (!stdout) return [];

  try {
    let rows = parsePlaylistJson(stdout);
    const brand = product.brand.trim();
    const sku = product.sku.trim();
    const productName = product.productName.trim();
    rows = rows.map((r) => ({
      ...r,
      brand,
      sku,
      productName,
      sourceQuery: qText,
    }));

    if (hasDate && dateRange) {
      const after = dateRange.after?.trim();
      const before = dateRange.before?.trim();
      rows = await filterRowsByUploadDate(
        rows,
        {
          after: after || undefined,
          before: before || undefined,
        },
        cap
      );
    } else {
      rows = rows.slice(0, cap);
    }
    return rows;
  } catch {
    return [];
  }
}
